app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+            {
+                keyFailure("Cannot setup wallet seed.");
+                return "";
+            }
+            Cipher outCipher = Cipher.getInstance(CIPHER_ALGORITHM);
+            final GCMParameterSpec spec = new GCMParameterSpec(128, iv);
+            outCipher.init(Cipher.DECRYPT_MODE, secretKey, spec);
+            CipherInputStream cipherInputStream = new CipherInputStream(encryptedHDKeyBytes, outCipher);
+            byte[] mnemonicBytes = readBytesFromStream(cipherInputStream);
+            String mnemonic = new String(mnemonicBytes);
+
+            switch (operation)
+            {
+                case FETCH_MNEMONIC:
+                    callbackInterface.FetchMnemonic(mnemonic);
+                    break;
+                case UPGRADE_HD_KEY:
+                case SIGN_DATA:
+                    return mnemonic;
+                default:
+                    break;
+            }
+        }
+        catch (InvalidKeyException e)
+        {
+            if (e instanceof UserNotAuthenticatedException)
+            {
+                checkAuthentication(operation);
+            }
+            else
+            {
+                keyFailure(e.getMessage());
+            }
+        }
+        catch (UnrecoverableKeyException e)
+        {
+            keyFailure("Key created at different security level. Please re-import key");
+            e.printStackTrace();
+        }
+        catch (IOException | CertificateException | KeyStoreException | NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException e)
+        {
+            e.printStackTrace();
+            keyFailure(e.getMessage());
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+        }
+
+        return "";
+    }
+
+    private FileInputStream getEncryptedFileStream(String currentKey) throws FileNotFoundException
+    {
+        String encryptedHDKeyPath = getEncryptedFilePath(currentKey);
+        if (encryptedHDKeyPath != null) return new FileInputStream(encryptedHDKeyPath);
+        else throw new FileNotFoundException("No key file found");
+    }
+
+    private void createHDKey()
+    {
+        HDWallet newWallet = new HDWallet(DEFAULT_KEY_STRENGTH, "");
+        storeHDKey(newWallet, Operation.CREATE_NON_AUTHENTICATED_KEY); //create non-authenticated key initially
+    }
+
+    /**
+     * Called after an authentication event was required, and the user has completed the authentication event
+     */
+    private void importHDKey()
+    {
+        //first recover the seed phrase from non-authlocked key. This removes the need to keep the seed phrase as a member on the heap
+        // - making the key operation more secure
+        String seedPhrase = unpackMnemonic(currentKeyAddress, UPGRADE_HD_KEY);
+        HDWallet newWallet = new HDWallet(seedPhrase, "");
+        storeHDKey(newWallet, Operation.IMPORT_HD_KEY);
+    }
+
+    private UpgradeKeyResult upgradeKey()
```

Comment:
Assuming the attacker has write access to the file system, there is a potential (non-severe) attack here: assume we have a non-auth-locked seed for address ADDR. The attacker can create a bogus file named ADDR-hd (note no no-auth label); in this case the if branch doesn't run, and instead the user is given to believe that the seed is already auth-locked. 

The first problem here is how `getEncryptedFilePath` is implemented (see that method for comments); the bigger problem is that file names are used as a source of truth. It would be better to tie checking if a key is auth-locked to the secure enclave instead.